KAFKA-6049: Add non-windowed Cogroup operator (KIP-150)

wcarlson5 opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
first commit on Cogroup operator

Commit:
first commit on Cogroup operator

wcarlson5 committed:
first commit on Cogroup operator

wcarlson5 committed:
first commit on Cogroup operator

wcarlson5 committed:
create cogroup for key value stores

wcarlson5 committed:
create cogroup for key value stores

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+        this.aggregateBuilder = new CogroupedStreamAggregateBuilder<>(builder);
+    }
+
+    @SuppressWarnings("unchecked")
```
it is necessary to  to do an unchecked cast for the input value type. This is because cogrouped can have any type value for the group streams it intakes

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+        final Collection<? extends AbstractStream<K, ?>> groupedStreams = new ArrayList<>(groupPatterns.keySet());
+        final AbstractStream<K, ?> kGrouped = groupedStreams.iterator().next();
+        groupedStreams.remove(kGrouped);
+        //kGrouped.ensureCopartitionWith(groupedStreams);
```
This will be added in a later PR

mjsax replied:
I think we should have this in this PR already.

wcarlson5 replied:
fair, I have added it back

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, AGGREGATE_NAME);
+        final ProcessorSupplier<K, Vout> tableSource = new KTableSource<>(
```
A new case will be added for windowed streams

wcarlson5 committed:
fixed indents

wcarlson5 committed:
fixed indents

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+ * @param <K> Type of keys
+ * @param <Vout> Type of values after agg
+ */
+public interface CogroupedKStream<K, Vout> {
```
Should it be `CoGroupedKStream`?

Should it be `VOut` ?

wcarlson5 replied:
I would say cogrouped is one word. But the VOut and VIn make sense

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+public interface CogroupedKStream<K, Vout> {
+
+    /**
+     * {@code CogroupedKStream} is an abstraction of multiple <i>grouped</i> record streams of
```
The JavaDoc does not describe what the method does. Maybe something like:
```
Add an already {@link KGroupedStream grouped KStream} to this {@code CoGroupedKStream}.
```

Btw: should we also update the JavaDocs of `KStream#groupByKey()` (and friends) to point out that a `KGroupedStream` can be cogrouped with others? Atm, those JavaDocs only talk about aggregation of a single KStream.

We should also describe repartitioning and Serdes? (cf. JavaDocs of `KStream#groupByKey()` and friends).

mjsax replied:
High level comment: I think we need to restructure the content a little different compared to single stream grouping/aggregation, because the methods are aligned differently. Not 100% sure how atm, thus, I might add more comments about this later on if I have new ideas.

wcarlson5 replied:
I added a line to the KStream JavaDocs, that was a good idea. I see your point to the fact that there are differences in the methods but they do essentially the same thing so I am not what other information is needed in this case.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     */
+    <Vin> CogroupedKStream<K, Vout> cogroup(final KGroupedStream<K, Vin> groupedStream,
+                                            final Aggregator<? super K, ? super Vin, Vout> aggregator);
+    /**
```
nit: add missing empty line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     * <p>
+     * It is an intermediate representation after a grouping of {@link KStream}s, before the
+     * aggregations are applied to the new partitions resulting in a {@link KTable}.
+     * <p>
```
Nit: we don't need this tag before the parameter list.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     * <p>
+     * @param groupedStream a group stream object
+     * @param aggregator   an {@link Aggregator} that computes a new aggregate result
+     * @param <Vin> Type of input values
```
Nit: should it be `VIn` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     * @param groupedStream a group stream object
+     * @param aggregator   an {@link Aggregator} that computes a new aggregate result
+     * @param <Vin> Type of input values
+     * @return a KCogroupedStream with the new grouopStream and aggregator linked
```
nit: `{@code KCogroupedStream}`
Typo `grouopStream`

What does "and aggregator linked" mean (I guess I can infer what you try to say, but I would just remove it)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+    /**
+     * Aggregate the values of records in these streams by the grouped key.
+     * Records with {@code null} key or value are ignored.
+     * Aggregating is a generalization of Reducing combining via
```
There is no `CoGroupedStream#reduce()` -- we can remove this

mjsax replied:
This was not addressed yet. the whole sentence can be removed

wcarlson5 replied:
addressed now

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     * The specified {@link Initializer} is applied once directly before the first input record is
+     * processed to provide an initial intermediate aggregation result that is used to process the
+     * first record.
+     * The specified {@link Aggregator} is applied for each input record and computes
```
This should go into `cogroup()` JavaDocs?

mjsax replied:
This method does not take an `Aggregator` as parameter...

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,187 @@
+     * For non-local keys, a custom RPC mechanism must be implemented using {@link
+     * KafkaStreams#allMetadata()} to query the value of the key on a parallel running instance of
+     * your Kafka Streams application.
+     *
```
nit: remove empty link

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -535,4 +535,22 @@
     SessionWindowedKStream<K, V> windowedBy(final SessionWindows windows);
 
+    /**
+     * {@code CogroupedKStream} is an abstraction of multiple <i>grouped</i> record streams of
```
As above. The JavaDoc should describe what the method does.
```
Create a new {@code CogroupedKStream} from the this grouped KStream to allow co-grouping other grouped KStreams to it.
```
(or similar)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+import org.apache.kafka.streams.kstream.internals.graph.StreamsGraphNode;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+public class CogroupedKStreamImpl<K, Vout> extends AbstractStream<K, Vout> implements CogroupedKStream<K, Vout> {
```
Should it be `CoGroupedKStreamImpl`?

Should it be `VOut` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+                         final Set<String> sourceNodes,
+                         final StreamsGraphNode streamsGraphNode,
+                         final InternalStreamsBuilder builder) {
+        super(name, null, null, sourceNodes, streamsGraphNode, builder);
```
Why do we not pass `Serdes` ? If we repartition, don't we need to have upstream Serdes to read from the repartition topic?

wcarlson5 replied:
Because it is not obvious from which stream to get the serde so either  override from the materialized object, or use the default from the config.

Also we do not pass the value Serdes because we cannot do so safely because each streams can have a different value type

mjsax replied:
Ah. I missed that we have only only `CogroupedKStreamImpl` object (my mental model was that we have one for each input stream).

What's unclear to me atm (maybe I need to do more detailed review) is, how repartitioning works? For that case, when do we insert a "source" node that is reading from the repartition topic, and where does the source node get the `Serde` information from? We could also have multiple independent repartition steps for different input streams.

wcarlson5 replied:
we had a couple long conversations about this. We resolved that the Serdes would come from the Materialized when aggregated or the default for that type

mjsax replied:
I think there is some misunderstanding. I understand that we cannot pass in `Serdes` to the superclass, but we still need to track them. Assume, the following programs:
```
KStream<String, Integer> stream1 = builder.stream(...).map(...);
KStream<String, Long> stream2 = builder.stream(...).map(...);

KTable<String, Float> result = stream1.groupByKey(Grouped.with(...))
    .cogroup(stream2.groupByKey(Grouped.with(...)))
    .aggregate(...);
```

Because we do a `map()` the serdes from `builder.stream()` cannot be forwarded and we also need to repartition each stream before we can compute the `aggregation()`. Hence, the use needs to pass in the Serdes for the new types (ie, result types of `map()` via `Grouped()` into each operation -- note, that both value types can be different. Because we apply repartitioning lazy, we would only insert it when we call `aggregate()`. Hence, the topology we need to generate later must be:
```
source1 -> map1 -> sink1 -> repartitionTopic1 -> source3 -> aggProcessor1 --+
                                                                            |
                                                                            + --> mergeNode
                                                                            |
source2 -> map2 -> sink2 -> repartitionTopic2 -> source4 -> aggProcessor2 --+
```

For both repartition steps, we need to use different Serdes (as provided by each `Grouped` parameter) -- hence, the `GroupedKStream` must forward the corresponding Serdes to `CogroupedKStreamImpl` because this class will add all corresponding `StreamGraphNodes` (including the ones to repartition the data if necessary. Compare https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java#L84-L99

Not sure atm what the best way is to implement this. Maybe just adding
```
final private Map<KGroupedStreamImpl<K, ?>, Serde> keySerde;
final private Map<KGroupedStreamImpl<K, ?>, Serde> valueSerde;
```

to `CogroupedKStreamImpl` (ie, next to `groupPatterns`) might be sufficient. Maybe there is a more elegant way? Maybe we can introduce a small helper Pojo to wrap `Aggregator` + both `Serdes` and have only one `Map` instead (to avoid scattering the state we track into different data structures)

The Serdes from `Materialized` only apply to the _output_ value type that can be different to the input value types (in fact, input value types can be different to each other, too).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -223,4 +224,14 @@
+            final Aggregator<? super K, ? super V, Vout> aggregator) {
+        Objects.requireNonNull(aggregator, "aggregator can't be null");
+        return new CogroupedKStreamImpl<K, Vout>(
+                name, sourceNodes,
```
nit: formatting -> single parameter per line (same next line)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+                                                   final Aggregator<? super K, ? super Vin, Vout> aggregator) {
+        Objects.requireNonNull(groupedStream, "groupedStream can't be null");
+        Objects.requireNonNull(aggregator, "aggregator can't be null");
+        groupPatterns.put((KGroupedStreamImpl<K, ?>) groupedStream, (Aggregator<? super K, ? super Object, Vout>) aggregator);
```
nit: line too long

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+        final NamedInternal named = NamedInternal.empty();
+        return doAggregate(initializer, named,
+                new MaterializedInternal<>(
+                        materialized, builder,
```
nit: single parameter per line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,91 @@
+    private KTable<K, Vout> doAggregate(final Initializer<Vout> initializer,
+                                        final NamedInternal named,
+                                        final MaterializedInternal<K, Vout, KeyValueStore<Bytes, byte[]>> materializedInternal) {
+        return this.aggregateBuilder.build(groupPatterns,
```
nit: move parameter to next line

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+        final KTable<String, String> customers = groupedOne.cogroup(STRING_AGGREGATOR)
+                .aggregate(STRING_INITIALIZER);
+
+        customers.toStream().process(processorSupplier);
```
It seem more natural to use `driver.createOutputTopic()` instead of the this "hack" to use `Processor` -- is there any reason why we can't use `TestOutputTopic` ?

wcarlson5 replied:
Not anymore. I was not sure how to make it work with time stamps, but I figured it out

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+        try (final TopologyTestDriver driver = new TopologyTestDriver(builder.build(), props)) {
+            final TestInputTopic<String, String> testInputTopic = driver.createInputTopic("one", new StringSerializer(), new StringSerializer());
+            testInputTopic.pipeInput("k1", "A", 0);
+            testInputTopic.pipeInput("k11", "A", 0);
```
nit: why not `k2` ?

Should we use `A`, `B`, `C`, `D` for the values to make it easier to understand the expected result? It's unclear which A is which below.

wcarlson5 replied:
fair

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+
+
+    @Test
+    public void shouldCogroupAndAggregateTwoKStreamsWithUniqueKeys() {
```
`WithDistinctKeys`?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+            testInputTopic.pipeInput("k1", "A", 0);
+            testInputTopic.pipeInput("k11", "A", 0);
+            testInputTopic.pipeInput("k11", "A", 0);
+            testInputTopic.pipeInput("k1", "A", 0);
```
Should we add different timestamps? Also include out-of-order data to check that output timestamp is computed correctly?

wcarlson5 replied:
There are other tests that use different timestamps and out of order. This is making sure it handles all cases

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+
+
+    @Test
+    public void shouldCogroupAndAggregateTwoKStreamsWithoutUniqueKeys() {
```
`WithOverlappingKeys` (or `WithSharedKeys`)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Test
+    public void testCogroupKeyMixedInTopicsandChageTypes() {
```
`shouldAllowDifferentOutputTypeInCoGroup`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Test
+    public void testCogroupKeyMixedInTopicsTypesandChageTypes() {
```
`shouldCoGroupStreamsWithDifferentInputTypes`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Test
+    public void testCogroupKeyMixedAggregators() {
```
Not sure what this test covers?

wcarlson5 replied:
it combines different aggregators

mjsax replied:
Well, `shouldCoGroupStreamsWithDifferentInputTypes` above uses two different aggregator already because the input type are different. Can we consider this test redundant?

wcarlson5 replied:
Yes and no. It does use 2 aggregators but they effectively do that same thing. This time since the produce different outputs it is testing that the aggregators were applied to the correct stream. Whereas the other case is get testing if different aggregators work. it might be a trivial distinction with quite a bit of overlap but I think that it serves a purpose.

mjsax reviewed (commented):
Did an initial pass. Mostly nits.

We need to have a test that trigger repartitioning though -- I think, atm repartitioning would not work correctly. This might require an integration test.

wcarlson5 committed:
updated javadocs and naming schemes

wcarlson5 committed:
updated javadocs and naming schemes

wcarlson5 committed:
updated test structure

wcarlson5 committed:
updated test structure

wcarlson5 commented:
retest this please

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added naming for processors

wcarlson5 committed:
added tests for cogrouped with three grouped streams

wcarlson5 committed:
added tests for cogrouped with three grouped streams

wcarlson5 commented:
retest this please

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+     * a new aggregate using the current aggregate (or for the very first record using the
+     * intermediate aggregation result provided via the {@link Initializer}) and the record's value.
+     * <p>
+     * Thus, {@code aggregate(Initializer, Aggregator, Materialized)} can be used to compute
```
Wrong signature of `aggregate(...)`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+     * The specified {@link Initializer} is applied once directly before the first input record is
+     * processed to provide an initial intermediate aggregation result that is used to process the
+     * first record.
+     * The specified {@link Aggregator} is applied for each input record and computes
```
This method does not take an `Aggregator` -- seems you c&p the JavaDocs from `KGroupedStream` without revisiting them?

wcarlson5 replied:
Well I defiantly missed a few thing when I revised them. I also forgot to update some of the docs after the discussion on the mailing list

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+     * <pre>
+     * KafkaStreams streams = ... // some aggregation on value type double
+     * String queryableStoreName = "storeName" // the store name should be the name of the store as defined by the Materialized instance
+     * ReadOnlyKeyValueStore<String, Long> localStore = streams.store(queryableStoreName, QueryableStoreTypes.<String, Long>keyValueStore());
```
This seems to an orthogonal issue: via KIP-258, this is still valid code, but it's also possible to the a `TimestampedKeyValueStore` -- would be good to update the existing JavaDocs in a separate PR against `trunk` and we can cherry-pick to `2.4` and `2.3` branches. -- If you don't want to do a PR, we should file a JIRA.

wcarlson5 replied:
I fixed the ones in this file. I will leave what to do after to you

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+     * @param initializer  an {@link Initializer} that computes an initial intermediate aggregation
+     *                     result
+     * @param materialized an instance of {@link Materialized} used to materialize a state store.
+     *                     Cannot be {@code null}.
```
Another side cleanup we could do in another PR (if file a JIRA) -- revisit all JavaDocs and use annotation about "cannot be null" or "can be null" -- note that `initializer` cannot be `null` either but we don't say anything about it.

mjsax replied:
Created https://issues.apache.org/jira/browse/KAFKA-9234

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+    /**
+     * Aggregate the values of records in these streams by the grouped key.
+     * Records with {@code null}
+     * key or value are ignored.
```
Why the line break?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,316 @@
+     * The changelog topic will be named "${applicationId}-${storeName}-changelog", where
+     * "applicationId" is user-specified in {@link StreamsConfig} via parameter {@link
+     * StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "storeName" is the provide store
+     * name defined in {@code Materialized}, and "-changelog" is a fixed suffix.
```
This overload does not take `Materialized` parameter

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -535,4 +535,23 @@
+     * <p>
+     * It is an intermediate representation after a grouping of {@link KStream}s, before the
+     * aggregations are applied to the new partitions resulting in a {@link KTable}.
+     * <p>
```
nit: `<p>` not needed because there is no following paragraph

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -535,4 +535,23 @@
+     * <p>
+     *
+     * @param aggregator an {@link Aggregator} that computes a new aggregate result
+    This conversation was marked as resolved by wcarlson5
```
`This conversation was marked as resolved by wcarlson5` :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -2028,6 +2028,7 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
     /**
      * Group the records by their current key into a {@link KGroupedStream} while preserving the original values
      * and default serializers and deserializers.
+     * GroupedStreams can be further grouped with other streams to form a cogroupedStream
```
Either use `Grouped streams` or `{@link KGroupedStream KGroupedStreams}`

`...to form a {@link CogroupedStream}.` (note, missing `.` at the end)

Same comment below.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -223,4 +224,16 @@
+                name,
+                sourceNodes,
+                streamsGraphNode,
+                builder).cogroup(this, aggregator);
```
Would it be more natural to pass in `this` and the `aggregator` into the constructor of `CogroupedKStreamImpl` instead of calling `.cogroup(...)`?

wcarlson5 replied:
well it is possible to have a cogroup with only one KGroupedStream in it. its unnecessary but makes logical sense

mjsax replied:
Sure, I understand that. That was not the question, but might not be too important anyway.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -0,0 +1,106 @@
+    private static final String AGGREGATE_NAME = "COGROUPKSTREAM-AGGREGATE-";
+
+    final private Map<KGroupedStreamImpl<K, ?>, Aggregator<? super K, ? super Object, VOut>> groupPatterns;
+    final private CogroupedStreamAggregateBuilder<K, VOut> aggregateBuilder;
```
it seem that `aggregateBuilder` is only used in `doAggregate()` -- hence, why do we need it as a member variable? Why not just create an instance in `doAggregate()` ?

wcarlson5 replied:
Because it will be used when adding the windowed PRs

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+
+class CogroupedStreamAggregateBuilder<K, VOut> {
+
+    private static final String AGGREGATE_NAME = "COGROUPKSTREAM-AGGREGATE-";
```
This is defined in `CogroupedKStreamImpl` and here -- seems to be redundant -- make the other one package private and reuse in this class instead.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+        final Collection<? extends AbstractStream<K, ?>> groupedStreams = new ArrayList<>(groupPatterns.keySet());
+        final AbstractStream<K, ?> kGrouped = groupedStreams.iterator().next();
+        groupedStreams.remove(kGrouped);
+        kGrouped.ensureCopartitionWith(groupedStreams);
```
It seems to be clumsy to get a "random" `AbstactStream` to call the method. Better make `ensureCopartitionWith()` a static method and pass in the full set of `groupedStreams` -- for the join case, you would pass in both `AbstractStream` that needs to be copartitioned.

wcarlson5 replied:
because we still need the build from the abstract stream we would still be choosing one at  "random" just inside the method.

mjsax replied:
Fair enough :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+            processors.add(statefulProcessorNode);
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, AGGREGATE_NAME);
```
`functionName` -- what is a "function"? maybe `processorName` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, AGGREGATE_NAME);
+        final ProcessorSupplier<K, VOut> tableSource = new KTableSource<>(
```
Why do we create a `KTableSource` ?

wcarlson5 replied:
It is being used to collect the results fo the different operations and input them into the table. And as far as I couple tell the KTableImpl needs it via the StatefulProcessorNode

wcarlson5 replied:
Changed to a Passthrough

mjsax replied:
`KTable` is just a logical abstraction. And there is already a `StateStore` that is shared by the processors that do the cogroup-aggregation -- hence, there is no need to have a second store. We just need to "merge" the output of the aggregation-processors together. Maybe you can reuse `KStreamPassThrough`?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+        final String functionName = named.orElseGenerateWithPrefix(builder, AGGREGATE_NAME);
+        final ProcessorSupplier<K, VOut> tableSource = new KTableSource<>(
+                        storeBuilder.name(),
+                        storeBuilder.name());
```
Why do we make the store always queryable?

wcarlson5 replied:
Is there a case we should not?

wcarlson5 replied:
when I don't specify a queryableName I get a Failed to flush state store error

wcarlson5 replied:
Changed to a Passthrough

mjsax replied:
This issue might resolve itself if you get rid of the `KTableSource` processor (the purpose of `KTableSource` is to model the case when a topic is consumed as table).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,154 @@
+            kStreamAggregate = new KStreamSessionWindowAggregate<>(sessionWindows, storeBuilder.name(), initializer, aggregator, sessionMerger);
+        } else {
+            throw new IllegalArgumentException(
+                    "must be a TimeWindowedStream or a SessionWindowedStream");
```
Could also be neither. Error message should be fixed

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,483 @@
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotHaveNullMaterMaterializedOnAggregate() {
+        cogroupedStream.aggregate(STRING_INITIALIZER, (Materialized<String, String, KeyValueStore<Bytes, byte[]>>) null);
```
`aggregate()` got more overload taking `Named` parameter -- please add corresponding NPE tests

wcarlson5 replied:
good idea

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,483 @@
+        final KGroupedStream<String, String> groupedOne = test.groupByKey();
+
+        final KTable<String, String> customers = groupedOne.cogroup(STRING_AGGREGATOR)
+                .aggregate(STRING_INITIALIZER, Named.as("test"));
```
Where do you validate that this name is picked up? -- Also, we we need to actually pipe input via TDD? Seems a `builder.build().describe()` would be sufficient to verify the name without the need to process any data?

mjsax replied:
With regard to naming: we should also check that the store in only queryable is a name is specified via `Materialized`.

mjsax replied:
I was using this test to print the topology and it shows two sub topologies while it should be one (seems the reason is that you use the same `StreamsBuilder` as in `setup()` method.

Also, the naming of the operators seems to be incorrect. Also wondering if `KGroupedStream#cogroup()` needs on overload that takes a `Named` parameter? Maybe not, but the specified `Named` from `aggregate()` would need to be used for other processors, too. Atm there is this weird `COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test`

```
Topologies:
   Sub-topology: 0
    Source: KSTREAM-SOURCE-0000000000 (topics: [topic])
      --> none

  Sub-topology: 1
    Source: KSTREAM-SOURCE-0000000001 (topics: [one])
      --> COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test
    Processor: COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000002])
      --> test
      <-- KSTREAM-SOURCE-0000000001
    Processor: test (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000002])
      --> KTABLE-TOSTREAM-0000000005
      <-- COGROUPKSTREAM-AGGREGATE-KSTREAM-SOURCE-0000000001test
    Processor: KTABLE-TOSTREAM-0000000005 (stores: [])
      --> KSTREAM-SINK-0000000006
      <-- test
    Sink: KSTREAM-SINK-0000000006 (topic: output)
      <-- KTABLE-TOSTREAM-0000000005
```

wcarlson5 replied:
Alright, the I changed the tests so that they only use the correct number of builders. And I discovered that orElseGenerateWithPrefix exists. so that should fix these problems

mjsax replied:
I still don't see that the test verifies the names? Again the question, why do we need to process data to verify if there right names are assigned?

wcarlson5 replied:
I added a test for this

mjsax reviewed (commented):
Sorry for the long wait... I'll stay in this one to hopefully merge by Wednesday.

Btw: Can you rebase your PR to `trunk` -- a PR was merged that changes how the gradle build works, and you should pick up this change.

wcarlson5 committed:
addressed comments on the PR

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+public interface CogroupedKStream<K, VOut> {
+
+    /**
+     * Add an already {@link KGroupedStream grouped KStream} to this {@code CoGroupedKStream}.
```
type `Co[g]roupedKStream`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+ * It is an intermediate representation of one or more {@link KStream}s in order to
+ * apply one or more aggregation operations on the original {@link KStream} records.
+ * <p>
+ * It is an intermediate representation after a grouping of {@link KStream}s, before the
```
`It is an intermediate representation` is somewhat redundant (use in the sentence before already. Maybe merge this sentence into the paragraph above?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * a new aggregate using the current aggregate (or for the very first record using the
+     * intermediate aggregation result provided via the {@link Initializer}) and the record's value.
+     * <p>
+     * Thus, {@code aggregate(Initializer, Materialized)} can be used to compute
```
use `{@link #aggregate(....)}` 

You would use `{@code }` only if you refer to the same class or same method (of void linking to yourself --- it's just a question of style obviously).

Also wondering: does this sentence actually describe what `aggregate()` does and thus is misplaced here?

wcarlson5 replied:
I does not necessarily have to be there

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * <p>
+     * Thus, {@code aggregate(Initializer, Materialized)} can be used to compute
+     * aggregate functions like count.
+     * It is an intermediate representation after a grouping of {@link KStream}s, before the
```
Why this sentence? The `cogroup()` method is not an intermediate representation of anything. Seems to be a mix up of class JavaDocs vs documenting what a method does?

wcarlson5 replied:
Yeah that was a duplicate of the one above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * StreamsConfig#CACHE_MAX_BYTES_BUFFERING_CONFIG cache size}, and {@link
+     * StreamsConfig#COMMIT_INTERVAL_MS_CONFIG commit intervall}.
+     * <p>
+     * To query the local {@link KeyValueStore} it must be obtained via {@link
```
`KeyValueStore` -> `TimestampedKeyValueStore`

wcarlson5 replied:
It does not require a TimestampedKeyValueStore

wcarlson5 replied:
https://issues.apache.org/jira/browse/KAFKA-9243

mjsax replied:
> It does not require a TimestampedKeyValueStore

`Materialized` does take `KeyValueStore<Bytes,byte[]>` but this byte-arrya-kv-store was wrapped by `KeyValueStore<K,V>` before but is now wrapped by a `TimestampedKeyValueStore<K,V>` -- it's not about the API of the method, but about documenting IQ correctly.

But we can also leave the JavaDocs as-is for now, as the show code will still work. Thanks for creating a ticket to fix all JavaDocs at once (this way, it will be at least consistent).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * <pre>
+     * KafkaStreams streams = ... // some aggregation on value type double
+     * String queryableStoreName = "storeName" // the store name should be the name of the store as defined by the Materialized instance
+     * TimestampedKeyValueStore<String, Long> localStore = streams.store(queryableStoreName, QueryableStoreTypes.<String, Long>keyValueStore());
```
`QueryableStoreTypes.<String, Long>timestampedKeyValueStore()`

wcarlson5 replied:
same as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * String queryableStoreName = "storeName" // the store name should be the name of the store as defined by the Materialized instance
+     * TimestampedKeyValueStore<String, Long> localStore = streams.store(queryableStoreName, QueryableStoreTypes.<String, Long>keyValueStore());
+     * String key = "some-key";
+     * Long aggForKey = localStore.get(key); // key must be local (application state is shared over all running Kafka Streams instances)
```
`get()` returns a `ValueAndTimestamp<Long>` -- not sure if we should add a `.value()` call to the right hand side or change the return type left hand side? Maybe easier to change to 
```
Long aggForKey = localStore.get(key).value(); // ...
```

wcarlson5 replied:
It actually uses a KeyValueStore. I have updated the docs to reflect that, I had missed a few places before .

wcarlson5 replied:
https://issues.apache.org/jira/browse/KAFKA-9243

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Aggregate the values of records in these streams by the grouped key.
+     * Records with {@code null} key or value are ignored.
+     * Aggregating is a generalization of Reducing combining via
```
Because there is not `CogroupedKStream#reduce()` method this sentence is not useful and should be removed.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * StreamsConfig#CACHE_MAX_BYTES_BUFFERING_CONFIG cache size}, and {@link
+     * StreamsConfig#COMMIT_INTERVAL_MS_CONFIG commit intervall}.
+     * <p>
+     * To query the local {@link KeyValueStore} it must be obtained via {@link
```
as above (more often below -- please fit all)

wcarlson5 replied:
Making a Ticket : https://issues.apache.org/jira/browse/KAFKA-9243

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -0,0 +1,295 @@
+     * The specified {@link Initializer} is applied once directly before the first input record is
+     * processed to provide an initial intermediate aggregation result that is used to process the
+     * first record.
+     * The specified {@link Named} is applied once to the processor combining the grouped streams.
```
Not sure what this means?

wcarlson5 replied:
clarified

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -535,4 +535,21 @@
     SessionWindowedKStream<K, V> windowedBy(final SessionWindows windows);
 
+    /**
+     * Create a new {@code CogroupedKStream} from the this grouped KStream to allow co-grouping other grouped KStreams to it.
```
type: `the this`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -535,4 +535,21 @@
+     * in order to apply one or more aggregation operations on the original {@link KGroupedStream}
+     * records.
+     * <p>
+     * It is an intermediate representation after a grouping of {@link KStream}s, before the
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,153 @@
+        this.builder = builder;
+    }
+
+    <KR, VIn, W extends Window> KTable<KR, VOut> build(
```
The output type is only `KTable<KR, VOut>` is we don't use windowing, right? Otherwise, it should be `KTable<KR, Windowed<VOut>>` ? Do we actually need two different `build()` methods for both cases? Maybe we can clean this up in the follow up PR though. Just something to think about

wcarlson5 replied:
We could split it into different builds but its not necessary right now. this pr doesn't make use of the W but ti will be used the later PRs

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -0,0 +1,494 @@
+
+
+    @Test
+    public void useNamed() {
```
nit: should this be something like `shouldNameProcessorsBasedOnNamedParameter` ?

wcarlson5 replied:
fair

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,153 @@
+            processors.add(statefulProcessorNode);
+            builder.addGraphNode(kGroupedStream.getKey().streamsGraphNode, statefulProcessorNode);
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
```
nit: why `functionName` ? To be more explicit, maybe `mergeProcessorName` ?

mjsax replied:
I reran the `useNamed()` test and the processors are now named `test-0` (for the aggregation processor) and `test` (for the merge processor) -- this will be hard for people to reason about. I would suggest to add more hard-coded additional information to get names like `test-cogroup-1` (or maybe `test-agg-1`?) and `test-merge` -- the names prefix that the user provides tires all those processors together and the user can identify that all those processor belong the `aggregate()` -- the additional suffix information helps to give more context what each processor does?

For `join` we do something similar (cf. `KStreamImplJoin#join()` method using `NamedInternal#suffixWithOrElseGet(...)`

wcarlson5 replied:
used suffixWithOrElseGet it fixed the problems I think

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -0,0 +1,153 @@
+        }
+        final String functionName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
+        final ProcessorSupplier<K, VOut> passThrough = new PassThrough<>();
+        final StatefulProcessorNode<K, VOut> tableSourceNode =
```
This is not a `tableSourceNode` and should be renamed. -- Also, why is this not a stateless processor?

wcarlson5 replied:
related to my earlier misunderstanding about table sources

mjsax reviewed (commented):
If you want, we can also split the PR into two, and add auto-repartitioning in a follow up PR. We also need tests for this case.

wcarlson5 committed:
addressed comments on the PR part 2

wcarlson5 committed:
addressed comments on the PR part 3

wcarlson5 committed:
remove extra import

mjsax committed:
Fix checkstyle error
Improved JavaDocs
Code reformatting
Added some more tests
Fixed naming and updated naming-test

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -65,224 +64,205 @@
      */
-    KTable<K, VOut> aggregate(final Initializer<VOut> initializer,
-                              final Materialized<K, VOut, KeyValueStore<Bytes, byte[]>> materialized);
+    KTable<K, VOut> aggregate(final Initializer<VOut> initializer);
```
I reordered the method from "few parameter" to "more parameters" to make it easier to navigate within the file.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -42,20 +41,20 @@
+     * streams of this {@code CogroupedKStream}.
+     * If this is not the case, you would need to call {@link KStream#through(String)} before
+     * {@link KStream#groupByKey() grouping} the {@link KStream}, using a pre-created topic with the "correct" number of
+     * partitions.
```
New paragraph.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -42,20 +41,20 @@
-     * {@link KeyValue} pairs.
-     * It is an intermediate representation of one or more {@link KStream}s
-     * in order to apply one or more aggregation operations on the original {@link KStream}
-     * records.
```
Removed this, as it describes `CogroupedKStream` what is not appropritate here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -65,224 +64,205 @@
-     * StreamsConfig#CACHE_MAX_BYTES_BUFFERING_CONFIG cache size}, and {@link
-     * StreamsConfig#COMMIT_INTERVAL_MS_CONFIG commit intervall}.
+     * To compute the aggregation the corresponding {@link Aggregator} as specified in
+     * {@link #cogroup(KGroupedStream, Aggregator) cogroup(...)} is used per input stream.
```
New sentence.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:
```diff
@@ -65,224 +64,205 @@
+     * To compute the aggregation the corresponding {@link Aggregator} as specified in
+     * {@link #cogroup(KGroupedStream, Aggregator) cogroup(...)} is used per input stream.
+     * The specified {@link Initializer} is applied once per key, directly before the first input record per key is
+     * processed to provide an initial intermediate aggregation result that is used to process the first record.
```
added `per key` (twice)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -536,16 +536,20 @@
+     * The specified {@link Aggregator} is applied in the actual {@link CogroupedKStream#aggregate(Initializer)
+     * aggregation} step for each input record and computes a new aggregate using the current aggregate (or for the very
+     * first record per key using the initial intermediate aggregation result provided via the {@link Initializer} that
+     * is passed into {@link CogroupedKStream#aggregate(Initializer)}) and the record's value.
```
New paragraph

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -31,9 +27,15 @@
 public class CogroupedKStreamImpl<K, VOut> extends AbstractStream<K, VOut> implements CogroupedKStream<K, VOut> {
 
     static final String AGGREGATE_NAME = "COGROUPKSTREAM-AGGREGATE-";
+    static final String MERGE_NAME = "COGROUPKSTREAM-MERGE-";
```
New default name for the merge-node

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -43,8 +45,8 @@
-        this.groupPatterns = new LinkedHashMap<>();
-        this.aggregateBuilder = new CogroupedStreamAggregateBuilder<>(builder);
+        groupPatterns = new LinkedHashMap<>();
+        aggregateBuilder = new CogroupedStreamAggregateBuilder<>(builder);
```
remove unnecessary `this.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImpl.java:
```diff
@@ -61,10 +63,7 @@
-        Objects.requireNonNull(materialized, "materialized can't be null");
-        final NamedInternal named = NamedInternal.empty();
-        return aggregate(initializer, named, materialized);
+        return aggregate(initializer, NamedInternal.empty(), materialized);
```
Unified the non-null check into a single place.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -31,90 +26,90 @@
-                    sessionMerger);
+                kGroupedStream.getValue(),
+                initializer,
+                named.suffixWithOrElseGet(
```
No need to pass in a `Named` -- we can just pass in the actual name as `String` directly -- otherwise we call `suffixWithOrElseGet` twice for no reason

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -31,90 +26,90 @@
         }
-        final String mergeProcessorName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
+        final String mergeProcessorName = named.suffixWithOrElseGet(
+            "-cogroup-merge",
```
Changed this to generate a name `<userName>-cogroup-merge` to align to `<userName>-cogroup-agg-<counter>` instead of just `<userName>` for the merge node.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/CogroupedStreamAggregateBuilder.java:
```diff
@@ -31,90 +26,90 @@
-            final SessionWindows sessionWindows,
-            final Merger<? super K, VOut> sessionMerger) {
-
-        final String processorName = named.orElseGenerateWithPrefix(builder, CogroupedKStreamImpl.AGGREGATE_NAME);
```
Removed this and pass in the `processorName` as `String` parameter directly.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -100,60 +94,102 @@ public void shouldNotHaveNullAggregatorOnCogroup() {
 
     @Test(expected = NullPointerException.class)
     public void shouldNotHaveNullInitializerOnAggregate() {
+        cogroupedStream.aggregate(null);
```
Added couple of more permutations for NPE tests.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java:
```diff
@@ -100,60 +94,102 @@ public void shouldNotHaveNullAggregatorOnCogroup() {
+        final KTable<String, String> customers = groupedOne
+            .cogroup(STRING_AGGREGATOR)
+            .cogroup(groupedTwo, STRING_AGGREGATOR)
+            .aggregate(STRING_INITIALIZER, Named.as("test"), Materialized.as("store"));
```
Also set the store name in this test.

mjsax commented:
@wcarlson5 I pushed a commit to fix jenkins checkstyle error -- will merge this after Jenkins is green.

We will need 4 follow up PRs:
 - add auto-repartitioning (including integration test)
 - time-windowed-cogroup
 - session-windowed-cogroup
 - update Scala API

The first three can be done in parallel IMHO. The last one only at the very end.
